alembic/script/revision.py

```diff
@@ -40,6 +40,38 @@ def __init__(self, message, argument):
         self.argument = argument
 
 
+class CycleDetected(RevisionError):
+    kind = "Cycle"
+
+    def __init__(self, revisions):
+        self.revisions = revisions
+        super(CycleDetected, self).__init__(
+            "%s is detected in revisions (%s)"
+            % (self.kind, ", ".join(revisions))
+        )
+
+
+class DependencyCycleDetected(CycleDetected):
+    kind = "Dependency cycle"
+
+    def __init__(self, revisions):
+        super(DependencyCycleDetected, self).__init__(revisions)
+
+
+class LoopDetected(CycleDetected):
+    kind = "Self-loop"
```

Comment:
**Koichiro Den** (lkpdn) wrote:

Done. PTAL at Patch Set 3.

*note: all involved members in a connected graph which contains one or more cycles are shown, but not exact members forming the cycles. If we want the latter to be shown, extra search would be required, which I think is not so worthy. So I kept it simple in Patch Set 3.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/2379